qemu/tests/balloon_check.py

```diff
@@ -91,6 +91,73 @@ def memory_check(self, step, ballooned_mem):
         return (mmem, gmem)
 
+    def enable_polling(self, device_path):
+        """
+        Enables polling in the specified interval
+
+        :param keyname: QOM path for the balloon device.
+        """
+        polling_interval = int(self.params.get("polling_interval", 2))
+        sleep_time = int(self.params.get("polling_sleep_time", 20))
+        error_context.context("Enable polling", logging.info)
+        self.vm.monitor.qom_set(device_path, "guest-stats-polling-interval", polling_interval)
+        time.sleep(sleep_time)
+
+    def get_stat(self, device_path):
+        """
+        Get memory statistics from qmp.
+
+        :param keyname: QOM path for the balloon device.
+        """
+        error_context.context("Get memory stat from qmp", logging.info)
+        return self.vm.monitor.qom_get(device_path, "guest-stats")
+
+    def memory_stats_compare(self, keyname, memory_stat_qmp):
+        """
+        Check whether memory statistics from qmp is same with guest memory.
+
+        :param keyname: key name of the output of the 'qom-get' property.
+        :param memory_stat_qmp: memory stat values from qmp.
+        """
+        check_mem_ratio = float(self.params.get("check_mem_ratio", 0.1))
+        error_context.context("Get memory from guest", logging.info)
+        if keyname == "stat-free-memory":
+            guest_mem = self.get_guest_free_mem(self.vm)
+        elif keyname == "stat-total-memory":
+            guest_mem = self.get_vm_mem(self.vm)
+        memory_stat_qmp = "%sB" % memory_stat_qmp
+        memory_stat_qmp = int(float(utils_misc.normalize_data_size(
+                                   memory_stat_qmp, order_magnitude="M")))
+        if (float(abs(guest_mem - memory_stat_qmp)) / guest_mem) > check_mem_ratio:
+            raise exceptions.TestError("%s of guest %s is not equal to %s in"
+                                       "qmp,the acceptable ratio is %s"
+                                       % (keyname, guest_mem, memory_stat_qmp,
+                                          check_mem_ratio))
+
+    def memory_stats_check(self, keyname):
+        """
+        Check whether memory statistics is working well via balloon service.
+
+        :param keyname: key name of the output of the 'qom-get' property.
+        :return: if memory stat can be got correctly via balloon service, return
+                 True, reture False when memory stat is not available.
+        :rtype: bool
+
+        """
+        base_path = self.params.get("base_path", "/machine/peripheral/")
+        device = self.params["balloon"]
+        device_path = base_path + device
+        mem_stat_disable = 0xffffffffffffffff
+
+        self.enable_polling(device_path)
+        memory_stat_qmp = self.get_stat(device_path)['stats'][keyname]
+
+        if memory_stat_qmp == mem_stat_disable:
+            return False
+        else:
+            self.memory_stats_compare(keyname, memory_stat_qmp)
```

Comment:
I think this step is not required by the function, you can call it independently after finished the current check.

Reply:
memory statistic work should based on statistic values are correct.
So this step should be a must for the function.

Reply:
Yes, memory_stats_compare() should do both step 1 and step 2.